pallets/pallet-bonded-coins/src/tests/curves/lmsr.rs

```diff
@@ -0,0 +1,133 @@
+use std::str::FromStr;
+
+use crate::{
+	curves::{BondingFunction, LMSRCalculation},
+	mock::{assert_relative_eq, Float},
+};
+
+#[test]
+fn mint_first_coin() {
+	// Create curve with liquidity parameter b=100_000_000, and passive issuance=0
+	let m = Float::from_num(100_000_000);
+	let passive_issuance = Float::from_num(0);
+
+	let curve = LMSRCalculation { m, passive_issuance };
+
+	let low = Float::from_num(0);
+	let high = Float::from_num(1);
+
+	// Costs for existing supply:  100000000 * ln(e^(0/100000000) + e^(0/100000000)) = 69314718.055994530942
+	// Costs for new supply: 100000000 * ln(e^(1/100000000) + e^(0/100000000)) = 69314718.555994532192
+	// Costs to mint the first coin: 69314718.555994532192 - 69314718.055994530942 = 0.50000000124972321215 -> 0.50000000124972321215
+	let costs = curve.calculate_costs(low, high).unwrap();
+
+	let expected_costs = Float::from_str("0.50000000124972321215").unwrap();
+
+	assert_relative_eq(costs, expected_costs, Float::from_str("0.00000002").unwrap());
+}
+
+#[test]
+fn high_supply_with_no_passive_issuance() {
+	// Create curve with liquidity parameter b=100_000_000, and passive issuance=0
+	let m = Float::from_num(100_000_000);
+	let passive_issuance = Float::from_num(0);
+	let curve = LMSRCalculation { m, passive_issuance };
+
+	// we mint 100 coins.
+	let low = Float::from_num(100_000_000u128);
+	let high = Float::from_num(100_000_100u128);
+
+	// Costs for existing supply:  100000000 * ln(e^(100000000/100000000) + e^(0/100000000)) = 131326168.7518222834
+	// Costs for new supply: 100000000 * ln(e^(100000100/100000000) + e^(0/100000000)) = 131326241.857689977
+	// Costs to mint the first coin: 131326241.857689977 - 131326168.7518222834 = 73.1058676936
+	let costs = curve.calculate_costs(low, high).unwrap();
+
+	let expected_costs = Float::from_str("73.1058676936").unwrap();
+
+	assert_relative_eq(costs, expected_costs, Float::from_str("0.000001").unwrap());
```

Comment:
It feels weird the epsilon changes with every unit test tho. What's the reason?

Reply:
I assume, intermediate results can not be fully represented. Depending on the input, the error might be bigger or smaller.